README.md
# Amazon Chime SDK for Browser (ES module)

### Build video calling, audio calling, and screen sharing applications powered by Amazon Chime.

This is a forked version of amazon chime sdk with es module support.
No need to build the package with webpack or similar bundler for using the original chime sdk (CommonJs) in browser


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.1] - 2020-05-20

### Added
- Add connectionDidBecomeGood callback in AudioVideoObserver
- Add an integration test for Data Message
- docs added

### Changed
- Styling and Markdown support for meeting demo chat
- package name changed
- git url changed
- more docs
- prebuild modified

### Removed

### Fixed
- Fix Firefox version 76 missing/grey tiles
- Fix data message integration tests
- Fix several integration test name

## [1.6.2] - 2020-05-18

### Fixed
- Disable audio capture for Electron Screen Capture

## [1.6.0] - 2020-05-15

### Added
- Allow option to skip device selection page in demo app.
- Add demo hook for debugging media connections
- Add github link to getNearestMediaRegion method in README.md
- Add data message APIs

### Changed
- Allow audio for screen capture in Chrome and Edge browsers
- Decouple the get call request from the UI
- Use getSettings if possible on MediaStream and move some info logs to debug level
- Use innerText instead of innerHTML

### Removed

### Fixed
- Fix minor coding styles for data message APIs

## [1.5.0] - 2020-05-07

### Added
- Add bandwidth policy to meeting session configuration to allow overriding default policies
- Add more content sharing integration tests
- Add gifs to read me file to show latest npm version and downloads
- Add method to get the nearest media region
- Display meeting and attendee IDs in the demo

### Changed
- Simplify meeting demos to leverage externalUserId in roster
- Update PR template to add testing information
- Support a mobile-friendly demo
- Increase the size of content share video tile for the demo app in small screen
- Update reconnection parameters in ConnectionHealthPolicyConfiguration

### Removed
- Remove unused VideoAdaptiveSubscribePolicy

### Fixed
- Fix serverless deploy script to work on Windows
- Clean up and fix serverless package bundling
- Do not mirror local video for rear-facing camera
- Fix sip url for meeting demo
- Fix local video freeze in Safari after toggling off and on
- Fix meeting demo content share turning off on attendee join
- Disable audio sample constraints when not using WebAudio
- Reset Sauce Lab session to make sure clean state
- Fix integration test emit metrics
- Fix the CloudWatch log handler

### Security
- Bump package-lock.json jquery to 3.5.0 and yargs-parser to 18.1.3

## [1.4.0] - 2020-04-24

### Added
- Expose an API for GetStats from RTCPeerConnection
- Add BrowserBehavior test for supported video codecs
- Expose ExternalUserID on videoTileDidUpdate

### Changed
- Use getByteTimeDomainData to support iOS Safari in meeting demo
- Update README to incorporate documentation feedback

### Removed

### Fixed
- Fix broken link in GitHub main page README

## [1.3.0] - 2020-04-17

### Added
- Enabled the use of send-side bandwidth estimation
- Add guide for content sharing
- Display meeting id in the demo app
- Add additional callback in AudioVideoObserver to indicate video downlink pressure
- Add meeting demo parameter for recording user
- Add a script demo to bundle Chime SDK into a single JS file
- Add device demo
- Add base infrastucture for demo app in react
- Add pricing link in README
- Add an overview of API methods
- Add IoT integration to device demo
- Add option to run integration tests locally
- Add the use case guide
- Upgrade dependency aws-iot-device-sdk version
- Add externalUserId to the tile properties
- Add post publish script
- Add feature flag to enable WebRTC Unified Plan for Chromium-based browsers
- Add link to Amazon Chime SDK Security in README

### Changed
- Prevent prebuild from increase patch number when publishing to NPM
- Change the cloudwatch log message format
- Only run integration tests if files on watchlist were modified
- Temporarily only run test in Chrome for Travis integration tests
- Allow content share frame rate to be configurable
- Move demo guides to demo folders
- Fix the default video resolution comment in DeviceController

### Removed
- Remove unimplemented callbacks remoteDidMuteAudio and remoteDidUnmuteAudio on AudioVideoObserver
- Remove the minimal demo app
- Removed incomplete demo and component directories

### Fixed
- Fix retry logic for integration test
- Update typedocs to 0.16 and re-generate doc files
- Fix issue in Travis script that prevents integration tests from running
- Fix markdown formatting with backticks in API overview
- Fix an issue that a dev dependnecy @types/dom-mediacapture-record is not getting installed
- Fix typo in README.md file
- Ensure that attendee presence leave and join are ordered correctly
- Fix video element issue in Iphone
- Fix post publish script to also include meeting v1
- Use build:publish for publish script
- Fix Travis deploy failure

## [1.2.1] - 2020-03-20

### Added
- Add BITRATES in SdkSignalFrame Type and regenerate corresponding JS and TS protocol files.
- Add new ContentShareController APIs
- Getting Started guide
- Doc guide generator
- Add basic component library setup
- Add injectable session URL rewrite function to support proxies
- Add POSTLogger for meeting sessions
- Integrate POSTLogger into the demo app
- Add content share integration test
- Enable POSTLogger for the serverless demo app
- Add max-content-share query parameter to allow 2 content share at the same time
- Add an integration test that checks only 2 content share are allowed

### Changed
- Add observer event for content sharing
- Stop content share if the media stream end
- Trap video sending SSRC change in two consecutive negotiation
- Do not bypass ice gathering based on sdp connection attributes for Safari on iOS
- Show SDK version in the demo meeting app
- Automatically patch a version for each commit

test/task/ReceiveTURNCredentialsTask.test.ts
// Copyright 2019-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as chai from 'chai';

import AudioVideoControllerState from '../../src/audiovideocontroller/AudioVideoControllerState';
import NoOpAudioVideoController from '../../src/audiovideocontroller/NoOpAudioVideoController';
import MeetingSessionConfiguration from '../../src/meetingsession/MeetingSessionConfiguration';
import MeetingSessionCredentials from '../../src/meetingsession/MeetingSessionCredentials';
import MeetingSessionStatusCode from '../../src/meetingsession/MeetingSessionStatusCode';
import MeetingSessionTURNCredentials from '../../src/meetingsession/MeetingSessionTURNCredentials';
import MeetingSessionURLs from '../../src/meetingsession/MeetingSessionURLs';
import TimeoutScheduler from '../../src/scheduler/TimeoutScheduler';
import ReceiveTURNCredentialsTask from '../../src/task/ReceiveTURNCredentialsTask';
import Task from '../../src/task/Task';
import DOMMockBehavior from '../dommock/DOMMockBehavior';
import DOMMockBuilder from '../dommock/DOMMockBuilder';

describe('ReceiveTURNCredentialsTask', () => {
  const expect: Chai.ExpectStatic = chai.expect;
  let domMockBuilder: DOMMockBuilder;
  let domMockBehavior: DOMMockBehavior;
  let context: AudioVideoControllerState;
  let task: Task;
  let testCredentials: MeetingSessionTURNCredentials;

  before(() => {
    testCredentials = new MeetingSessionTURNCredentials();
    testCredentials.username = 'fakeUsername';
    testCredentials.password = 'fakeTURNCredentials';
    testCredentials.ttl = Infinity;
    testCredentials.uris = ['fakeUDPURI', 'fakeTCPURI'];
  });

  beforeEach(() => {
    domMockBehavior = new DOMMockBehavior();
    domMockBuilder = new DOMMockBuilder(domMockBehavior);
    context = new AudioVideoControllerState();
    context.audioVideoController = new NoOpAudioVideoController();
    context.logger = context.audioVideoController.logger;
    context.meetingSessionConfiguration = new MeetingSessionConfiguration();
    context.meetingSessionConfiguration.urls = new MeetingSessionURLs();
    context.meetingSessionConfiguration.urls.turnControlURL = 'http://example.com';
    context.meetingSessionConfiguration.meetingId = 'testId';
    context.meetingSessionConfiguration.credentials = new MeetingSessionCredentials();
    context.meetingSessionConfiguration.credentials.joinToken = 'testToken';
    task = new ReceiveTURNCredentialsTask(context);
  });

  afterEach(() => {
    domMockBuilder.cleanup();
  });

  describe('run', () => {
    it('can get TURN credentials', async () => {
      await task.run();
      expect(context.turnCredentials).to.deep.equal(testCredentials);
    });

    it('handles when the fetch call fails', done => {
      domMockBehavior.fetchSucceeds = false;
      task
        .run()
        .then(() => {
          done(new Error('This line should not be reached.'));
        })
        .catch(() => {
          expect(context.turnCredentials).to.equal(null);
          done();
        });
    });

    it('handles a 403 error to fetch TURN Credentials', async () => {
      domMockBehavior.fetchSucceeds = true;
      domMockBehavior.responseStatusCode = 403;
      domMockBehavior.responseSuccess = false;
      try {
        await task.run();
        throw new Error('This line should not be reached.');
      } catch (error) {
        expect(error.message).includes(
          `the meeting status code: ${MeetingSessionStatusCode.ForbiddenToFetchTurnCredentials}`
        );
      }
    });

    it('will bypass the task when a url is not specified', async () => {
      context.meetingSessionConfiguration.urls.turnControlURL = null;
      task = new ReceiveTURNCredentialsTask(context);
      await task.run();
      expect(context.turnCredentials).to.equal(null);
    });
  });

  describe('cancel', () => {
    it('cancels a task when the session is timed out', done => {
      let called = false;

      domMockBehavior.asyncWaitMs = 500;
      new TimeoutScheduler(50).start(() => {
        task.cancel();
      });

      task
        .run()
        .then(() => {
          done(new Error('This line should not be reached.'));
        })
        .catch(() => {
          called = true;
        });

      new TimeoutScheduler(domMockBehavior.asyncWaitMs + 50).start(() => {
        expect(called).to.be.true;
        expect(context.turnCredentials).to.equal(null);
        done();
      });
    });
  });
});


test/meetingsession/MeetingSessionStatus.test.ts
// Copyright 2019-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as chai from 'chai';

import MeetingSessionStatus from '../../src/meetingsession/MeetingSessionStatus';
import MeetingSessionStatusCode from '../../src/meetingsession/MeetingSessionStatusCode';
import {
  SdkAudioStatusFrame,
  SdkErrorFrame,
  SdkSignalFrame,
} from '../../src/signalingprotocol/SignalingProtocol.js';

describe('MeetingSessionStatus', () => {
  const expect: Chai.ExpectStatic = chai.expect;

  const statusCodes = [
    MeetingSessionStatusCode.OK,
    MeetingSessionStatusCode.Left,
    MeetingSessionStatusCode.AudioJoinedFromAnotherDevice,
    MeetingSessionStatusCode.AudioDisconnectAudio,
    MeetingSessionStatusCode.AudioAuthenticationRejected,
    MeetingSessionStatusCode.AudioCallAtCapacity,
    MeetingSessionStatusCode.AudioCallEnded,
    MeetingSessionStatusCode.AudioInternalServerError,
    MeetingSessionStatusCode.AudioServiceUnavailable,
    MeetingSessionStatusCode.AudioDisconnected,
    MeetingSessionStatusCode.ForbiddenToFetchTurnCredentials,
    MeetingSessionStatusCode.VideoCallSwitchToViewOnly,
    MeetingSessionStatusCode.VideoCallAtSourceCapacity,
    MeetingSessionStatusCode.SignalingBadRequest,
    MeetingSessionStatusCode.SignalingInternalServerError,
    MeetingSessionStatusCode.SignalingRequestFailed,
    MeetingSessionStatusCode.StateMachineTransitionFailed,
    MeetingSessionStatusCode.ICEGatheringTimeoutWorkaround,
    MeetingSessionStatusCode.ConnectionHealthReconnect,
    MeetingSessionStatusCode.RealtimeApiFailed,
  ];

  describe('isFailure', () => {
    it('is a failure', () => {
      const failureSet = new Set([
        MeetingSessionStatusCode.AudioAuthenticationRejected,
        MeetingSessionStatusCode.AudioCallAtCapacity,
        MeetingSessionStatusCode.AudioInternalServerError,
        MeetingSessionStatusCode.AudioServiceUnavailable,
        MeetingSessionStatusCode.AudioDisconnected,
        MeetingSessionStatusCode.VideoCallAtSourceCapacity,
        MeetingSessionStatusCode.SignalingBadRequest,
        MeetingSessionStatusCode.SignalingInternalServerError,
        MeetingSessionStatusCode.SignalingRequestFailed,
        MeetingSessionStatusCode.StateMachineTransitionFailed,
        MeetingSessionStatusCode.ICEGatheringTimeoutWorkaround,
        MeetingSessionStatusCode.ConnectionHealthReconnect,
        MeetingSessionStatusCode.RealtimeApiFailed,
      ]);
      for (const statusCode of statusCodes) {
        const status = new MeetingSessionStatus(statusCode);
        expect(status.isFailure()).to.equal(failureSet.has(status.statusCode()));
      }
    });
  });

  describe('isTerminal', () => {
    it('is terminal', () => {
      const terminalSet = new Set([
        MeetingSessionStatusCode.Left,
        MeetingSessionStatusCode.AudioJoinedFromAnotherDevice,
        MeetingSessionStatusCode.AudioAuthenticationRejected,
        MeetingSessionStatusCode.AudioCallAtCapacity,
        MeetingSessionStatusCode.AudioCallEnded,
        MeetingSessionStatusCode.AudioInternalServerError,
        MeetingSessionStatusCode.AudioDisconnected,
        MeetingSessionStatusCode.ForbiddenToFetchTurnCredentials,
        MeetingSessionStatusCode.SignalingBadRequest,
        MeetingSessionStatusCode.SignalingInternalServerError,
        MeetingSessionStatusCode.SignalingRequestFailed,
        MeetingSessionStatusCode.VideoCallAtSourceCapacity,
        MeetingSessionStatusCode.RealtimeApiFailed,
      ]);
      for (const statusCode of statusCodes) {
        const status = new MeetingSessionStatus(statusCode);
        expect(status.isTerminal()).to.equal(terminalSet.has(status.statusCode()));
      }
    });
  });

  describe('isAudioConnectionFailure', () => {
    it('is an audio connection failure', () => {
      const audioConnectionFailureSet = new Set([
        MeetingSessionStatusCode.AudioAuthenticationRejected,
        MeetingSessionStatusCode.AudioInternalServerError,
        MeetingSessionStatusCode.AudioServiceUnavailable,
        MeetingSessionStatusCode.StateMachineTransitionFailed,
        MeetingSessionStatusCode.ICEGatheringTimeoutWorkaround,
        MeetingSessionStatusCode.SignalingBadRequest,
        MeetingSessionStatusCode.SignalingInternalServerError,
        MeetingSessionStatusCode.SignalingRequestFailed,
        MeetingSessionStatusCode.RealtimeApiFailed,
      ]);
      for (const statusCode of statusCodes) {
        const status = new MeetingSessionStatus(statusCode);
        expect(status.isAudioConnectionFailure()).to.equal(
          audioConnectionFailureSet.has(status.statusCode())
        );
      }
    });
  });

  describe('fromSignalFrame', () => {
    it('handles errors', () => {
      const errorStatusCodes: { [code: number]: MeetingSessionStatusCode } = {
        206: MeetingSessionStatusCode.VideoCallSwitchToViewOnly,
        509: MeetingSessionStatusCode.VideoCallAtSourceCapacity,
        200: MeetingSessionStatusCode.OK,
        300: MeetingSessionStatusCode.SignalingRequestFailed,
        400: MeetingSessionStatusCode.SignalingBadRequest,
        500: MeetingSessionStatusCode.SignalingInternalServerError,
      };

      for (const errorStatusCode in errorStatusCodes) {
        const message = SdkSignalFrame.create();
        message.type = SdkSignalFrame.Type.JOIN;
        message.error = SdkErrorFrame.create();
        message.error.status = Number(errorStatusCode);
        const status = MeetingSessionStatus.fromSignalFrame(message);
        expect(status.statusCode()).to.equal(errorStatusCodes[errorStatusCode]);
      }
    });

    it('handles audio frames', () => {
      const audioStatusCodes: { [code: number]: MeetingSessionStatusCode } = {
        200: MeetingSessionStatusCode.OK,
        301: MeetingSessionStatusCode.AudioJoinedFromAnotherDevice,
        302: MeetingSessionStatusCode.AudioDisconnectAudio,
        403: MeetingSessionStatusCode.AudioAuthenticationRejected,
        409: MeetingSessionStatusCode.AudioCallAtCapacity,
        410: MeetingSessionStatusCode.AudioCallEnded,
        500: MeetingSessionStatusCode.AudioInternalServerError,
        503: MeetingSessionStatusCode.AudioServiceUnavailable,
        201: MeetingSessionStatusCode.OK,
        400: MeetingSessionStatusCode.AudioDisconnected,
      };

      for (const audioStatusCode in audioStatusCodes) {
        const message = SdkSignalFrame.create();
        message.type = SdkSignalFrame.Type.AUDIO_STATUS;
        message.audioStatus = SdkAudioStatusFrame.create();
        message.audioStatus.audioStatus = Number(audioStatusCode);
        const status = MeetingSessionStatus.fromSignalFrame(message);
        expect(status.statusCode()).to.equal(audioStatusCodes[audioStatusCode]);
      }
    });

    it('handles audio frames but without audioStatus', () => {
      const message = SdkSignalFrame.create();
      message.type = SdkSignalFrame.Type.AUDIO_STATUS;
      const status = MeetingSessionStatus.fromSignalFrame(message);
      expect(status.statusCode()).to.equal(MeetingSessionStatusCode.SignalingRequestFailed);
    });

    it('handles non-error or non-audio frames', () => {
      const message = SdkSignalFrame.create();
      message.type = SdkSignalFrame.Type.JOIN;
      const status = MeetingSessionStatus.fromSignalFrame(message);
      expect(status.statusCode()).to.equal(MeetingSessionStatusCode.OK);
    });
  });
});


src/versioning/Versioning.ts
// Copyright 2019-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import DefaultBrowserBehavior from '../browserbehavior/DefaultBrowserBehavior';

export default class Versioning {
  static X_AMZN_VERSION = 'X-Amzn-Version';
  static X_AMZN_USER_AGENT = 'X-Amzn-User-Agent';

  /**
   * Return string representation of SDK version
   */
  static get sdkVersion(): string {
    return '1.0.2';
  }

  /**
   * Return low-resolution string representation of SDK user agent (e.g. `chrome-78`)
   */
  static get sdkUserAgentLowResolution(): string {
    const browserBehavior = new DefaultBrowserBehavior();
    return `${browserBehavior.name()}-${browserBehavior.majorVersion()}`;
  }

  /**
   * Return URL with versioning information appended
   */
  static urlWithVersion(url: string): string {
    const urlWithVersion = new URL(url);
    urlWithVersion.searchParams.append(Versioning.X_AMZN_VERSION, Versioning.sdkVersion);
    urlWithVersion.searchParams.append(
      Versioning.X_AMZN_USER_AGENT,
      Versioning.sdkUserAgentLowResolution
    );
    return urlWithVersion.toString();
  }
}


src/task/ReceiveTURNCredentialsTask.ts
// Copyright 2019-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import AudioVideoControllerState from '../audiovideocontroller/AudioVideoControllerState';
import ContentShareConstants from '../contentsharecontroller/ContentShareConstants';
import MeetingSessionStatusCode from '../meetingsession/MeetingSessionStatusCode';
import MeetingSessionTURNCredentials from '../meetingsession/MeetingSessionTURNCredentials';
import Versioning from '../versioning/Versioning';
import BaseTask from './BaseTask';

/*
 * [[ReceiveTURNCredentialsTask]] asynchronously retrieves TURN credentials.
 */
export default class ReceiveTURNCredentialsTask extends BaseTask {
  protected taskName = 'ReceiveTURNCredentialsTask';

  private url: string;
  private meetingId: string;
  private joinToken: string;
  private cancelPromise: (error: Error) => void;

  constructor(private context: AudioVideoControllerState) {
    super(context.logger);
    this.url = context.meetingSessionConfiguration.urls.turnControlURL;
    this.meetingId = context.meetingSessionConfiguration.meetingId;
    this.joinToken = context.meetingSessionConfiguration.credentials.joinToken;
  }

  cancel(): void {
    const error = new Error(`canceling ${this.name()}`);
    this.cancelPromise && this.cancelPromise(error);
  }

  async run(): Promise<void> {
    if (!this.url) {
      this.context.logger.info('skipping TURN credentials');
      return;
    }

    const options: RequestInit = {
      method: 'POST',
      mode: 'cors',
      cache: 'no-cache',
      credentials: 'omit',
      headers: {
        'Content-Type': 'application/json',
        'X-Chime-Auth-Token':
          '_aws_wt_session=' + this.joinToken.replace(ContentShareConstants.Modality, ''),
      },
      redirect: 'follow',
      referrer: 'no-referrer',
      body: JSON.stringify({ meetingId: this.meetingId }),
    };

    this.context.logger.info(`requesting TURN credentials from ${this.url}`);

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const responseBodyJson = await new Promise<any>(async (resolve, reject) => {
      this.cancelPromise = (error: Error) => {
        reject(error);
      };

      try {
        const responseBody = await fetch(Versioning.urlWithVersion(this.url), options);
        this.context.logger.info(`received TURN credentials`);
        if (responseBody.status && responseBody.status === 403) {
          let error = new Error(
            `canceling ${this.name()} due to the meeting status code: ${
              MeetingSessionStatusCode.ForbiddenToFetchTurnCredentials
            }`
          );
          reject(error);
        }
        resolve(await responseBody.json());
      } catch (error) {
        reject(error);
      }
    });

    this.context.turnCredentials = new MeetingSessionTURNCredentials();
    this.context.turnCredentials.password = responseBodyJson.password;
    this.context.turnCredentials.ttl = responseBodyJson.ttl;
    this.context.turnCredentials.uris = responseBodyJson.uris.map((uri: string): string => {
      return this.context.meetingSessionConfiguration.urls.urlRewriter(uri);
    });
    this.context.turnCredentials.username = responseBodyJson.username;
  }
}


src/meetingsession/MeetingSessionStatusCode.ts
// Copyright 2019-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export enum MeetingSessionStatusCode {
  /**
   * Everything is OK so far.
   */
  OK = 0,

  /**
   * The attendee left the meeting normally.
   */
  Left = 1,

  /**
   * The attendee joined from another device.
   */
  AudioJoinedFromAnotherDevice = 2,

  /**
   * The attendee should explicitly switch itself from joined with audio to
   * checked-in.
   */
  AudioDisconnectAudio = 3,

  /**
   * Authentication was rejected. The client is not allowed on this call.
   */
  AudioAuthenticationRejected = 4,

  /**
   * The client can not join because the call is at capacity.
   */
  AudioCallAtCapacity = 5,

  /**
   * The call was ended.
   */
  AudioCallEnded = 6,

  /**
   * There was an internal server error with the audio leg.
   */
  AudioInternalServerError = 7,

  /**
   * Could not connect the audio leg due to the service being unavailable.
   */
  AudioServiceUnavailable = 8,

  /**
   * The audio leg failed.
   */
  AudioDisconnected = 9,

  /**
   * The client has asked to send and receive video, but it is only possible to
   * continue in view-only mode (receiving video). This should be handled by
   * explicitly switching to view-only mode.
   */
  VideoCallSwitchToViewOnly = 10,

  /** This can happen when you attempt to join a video call in "send only" mode
  (transmitting your camera, but not receiving anything -- this isn't something
  we ever do in practice, but it is supported on the server). It should be
  treated as "fatal" and probably should not be retried (despite the 5xx nature). */
  VideoCallAtSourceCapacity = 11,

  /**
   * Bad request on JOIN or SUBSCRIBE
   */
  SignalingBadRequest = 12,

  /**
   * Internal server error on JOIN or SUBSCRIBE
   */
  SignalingInternalServerError = 13,

  /**
   * Received unknown signaling error frame
   */
  SignalingRequestFailed = 14,

  /**
   * Failed to transition between two states for some reason
   */
  StateMachineTransitionFailed = 15,

  /**
   * Timed out gathering ICE candidates. If in Chrome, this could be an
   * indication that the browser is in a bad state due to a VPN reconnect and
   * the user should try quitting and relaunching the app. See:
   * https://bugs.chromium.org/p/webrtc/issues/detail?id=9097
   */
  ICEGatheringTimeoutWorkaround = 16,

  /**
   * Due to connection health, a reconnect has been triggered.
   */
  ConnectionHealthReconnect = 17,

  /**
   * The realtime API failed in some way. This indicates a fatal problem.
   */
  RealtimeApiFailed = 18,

  /**
   * A task failed for an unknown reason.
   */
  TaskFailed = 19,

  /**
   * Audio device has switched.
   */
  AudioDeviceSwitched = 20,

  /**
   * Session update produces incompatible SDP.
   */
  IncompatibleSDP = 21,

  /**
   * This can happen when you attempt to join a meeting which has ended
   */
  ForbiddenToFetchTurnCredentials = 22,
}

export default MeetingSessionStatusCode;


src/meetingsession/MeetingSessionStatus.ts
// Copyright 2019-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { SdkSignalFrame } from '../signalingprotocol/SignalingProtocol.js';
import MeetingSessionStatusCode from './MeetingSessionStatusCode';

/**
 * [[MeetingSessionStatus]] indicates a status received regarding the session.
 */
export default class MeetingSessionStatus {
  constructor(private _statusCode: MeetingSessionStatusCode) {}

  statusCode(): MeetingSessionStatusCode {
    return this._statusCode;
  }

  isFailure(): boolean {
    switch (this._statusCode) {
      case MeetingSessionStatusCode.AudioAuthenticationRejected:
      case MeetingSessionStatusCode.AudioCallAtCapacity:
      case MeetingSessionStatusCode.AudioInternalServerError:
      case MeetingSessionStatusCode.AudioServiceUnavailable:
      case MeetingSessionStatusCode.AudioDisconnected:
      case MeetingSessionStatusCode.VideoCallAtSourceCapacity:
      case MeetingSessionStatusCode.SignalingBadRequest:
      case MeetingSessionStatusCode.SignalingInternalServerError:
      case MeetingSessionStatusCode.SignalingRequestFailed:
      case MeetingSessionStatusCode.StateMachineTransitionFailed:
      case MeetingSessionStatusCode.ICEGatheringTimeoutWorkaround:
      case MeetingSessionStatusCode.ConnectionHealthReconnect:
      case MeetingSessionStatusCode.RealtimeApiFailed:
      case MeetingSessionStatusCode.TaskFailed:
        return true;
      default:
        return false;
    }
  }

  isTerminal(): boolean {
    switch (this._statusCode) {
      case MeetingSessionStatusCode.Left:
      case MeetingSessionStatusCode.AudioJoinedFromAnotherDevice:
      case MeetingSessionStatusCode.AudioAuthenticationRejected:
      case MeetingSessionStatusCode.AudioCallAtCapacity:
      case MeetingSessionStatusCode.AudioCallEnded:
      case MeetingSessionStatusCode.AudioInternalServerError:
      case MeetingSessionStatusCode.AudioDisconnected:
      case MeetingSessionStatusCode.ForbiddenToFetchTurnCredentials:
      case MeetingSessionStatusCode.SignalingBadRequest:
      case MeetingSessionStatusCode.SignalingInternalServerError:
      case MeetingSessionStatusCode.SignalingRequestFailed:
      case MeetingSessionStatusCode.VideoCallAtSourceCapacity:
      case MeetingSessionStatusCode.RealtimeApiFailed:
        return true;
      default:
        return false;
    }
  }

  isAudioConnectionFailure(): boolean {
    switch (this._statusCode) {
      case MeetingSessionStatusCode.AudioAuthenticationRejected:
      case MeetingSessionStatusCode.AudioInternalServerError:
      case MeetingSessionStatusCode.AudioServiceUnavailable:
      case MeetingSessionStatusCode.StateMachineTransitionFailed:
      case MeetingSessionStatusCode.ICEGatheringTimeoutWorkaround:
      case MeetingSessionStatusCode.SignalingBadRequest:
      case MeetingSessionStatusCode.SignalingInternalServerError:
      case MeetingSessionStatusCode.SignalingRequestFailed:
      case MeetingSessionStatusCode.RealtimeApiFailed:
        return true;
      default:
        return false;
    }
  }

  static fromSignalFrame(frame: SdkSignalFrame): MeetingSessionStatus {
    if (frame.error && frame.error.status) {
      return this.fromSignalingStatus(frame.error.status);
    } else if (frame.type === SdkSignalFrame.Type.AUDIO_STATUS) {
      if (frame.audioStatus) {
        return this.fromAudioStatus(frame.audioStatus.audioStatus);
      }
      return new MeetingSessionStatus(MeetingSessionStatusCode.SignalingRequestFailed);
    }
    return new MeetingSessionStatus(MeetingSessionStatusCode.OK);
  }

  private static fromAudioStatus(status: number): MeetingSessionStatus {
    // TODO: Add these numbers to proto definition and reference them here.
    switch (status) {
      case 200:
        return new MeetingSessionStatus(MeetingSessionStatusCode.OK);
      case 301:
        return new MeetingSessionStatus(MeetingSessionStatusCode.AudioJoinedFromAnotherDevice);
      case 302:
        return new MeetingSessionStatus(MeetingSessionStatusCode.AudioDisconnectAudio);
      case 403:
        return new MeetingSessionStatus(MeetingSessionStatusCode.AudioAuthenticationRejected);
      case 409:
        return new MeetingSessionStatus(MeetingSessionStatusCode.AudioCallAtCapacity);
      case 410:
        return new MeetingSessionStatus(MeetingSessionStatusCode.AudioCallEnded);
      case 500:
        return new MeetingSessionStatus(MeetingSessionStatusCode.AudioInternalServerError);
      case 503:
        return new MeetingSessionStatus(MeetingSessionStatusCode.AudioServiceUnavailable);
      default:
        switch (Math.floor(status / 100)) {
          case 2:
            return new MeetingSessionStatus(MeetingSessionStatusCode.OK);
          default:
            return new MeetingSessionStatus(MeetingSessionStatusCode.AudioDisconnected);
        }
    }
  }

  private static fromSignalingStatus(status: number): MeetingSessionStatus {
    // TODO: Add these numbers to proto definition and reference them here.
    switch (status) {
      case 206:
        return new MeetingSessionStatus(MeetingSessionStatusCode.VideoCallSwitchToViewOnly);
      case 509:
        return new MeetingSessionStatus(MeetingSessionStatusCode.VideoCallAtSourceCapacity);
      default:
        switch (Math.floor(status / 100)) {
          case 2:
            return new MeetingSessionStatus(MeetingSessionStatusCode.OK);
          case 4:
            return new MeetingSessionStatus(MeetingSessionStatusCode.SignalingBadRequest);
          case 5:
            return new MeetingSessionStatus(MeetingSessionStatusCode.SignalingInternalServerError);
          default:
            return new MeetingSessionStatus(MeetingSessionStatusCode.SignalingRequestFailed);
        }
    }
  }
}
